docs/root/configuration/http/http_filters/decompressor_filter.rst

```diff
@@ -0,0 +1,64 @@
+.. _config_http_filters_decompressor:
+
+Decompressor
+============
+Decompressor is an HTTP filter which enables Envoy to bidirectionally decompress data.
+
+
+Configuration
+-------------
+* :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.decompressor.v3.Decompressor>`
+
+How it works
+------------
+When the decompressor filter is enabled, headers are inspected to
+determine whether or not the content should be decompressed. The content is
+decompressed and passed on to the rest of the filter chain. Note that decompression happens
+independently for request and responses based on the rules described below.
+
+Currently the filter supports :ref:`gzip compression <envoy_v3_api_msg_extensions.compression.gzip.decompressor.v3.Gzip>`
+only. Other compression libraries can be supported as extensions.
+
+An example configuration of the filter may look like the following:
+
+.. code-block:: yaml
+
+    http_filters:
+    - name: decompressor
+      typed_config:
+        "@type": type.googleapis.com/envoy.extensions.filters.http.decompressor.v3.Decompressor
+        decompressor_library:
+          name: basic
+          typed_config:
+            "@type": type.googleapis.com/envoy.extensions.compression.gzip.decompressor.v3.Gzip
+            window_bits: 10
+
+By *default* decompression will be *skipped* when:
+
+- A request/response does NOT contain *content-encoding* header.
+- A request/response includes *content-encoding* header, but it does not contain the configured
+  decompressor's content-encoding.
+- A request/response contains a *cache-control* header whose value includes "no-transform".
+
+When decompression is *applied*:
+
+- The *content-length* is removed from headers.
```

Comment:
Is this going to break some uses cases? Is it a future work item that we might want to buffer the decompressed data and rewrite content-length?

Reply:
Ah, sorry @mattklein123 , this context got dropped because github didn't allow me to move my original PR over.

We discussed here: https://github.com/junr03/envoy/pull/1#discussion_r411669576

And then I had this idea: https://github.com/envoyproxy/envoy/pull/11172#issuecomment-628949673

tl;dr: I originally implemented buffering here. But then decided to remove (and depend on the buffer filter) to keep this filter simpler. wdyt?

Reply:
Does the buffer filter reset content length? Anyway I think that's fine but I might make that clear in the documentation that the user might want to couple this with the buffer filter.

Reply:
> Does the buffer filter reset content length?

Yep, it sets it if not present, which made it ideal for this arrangement.

https://github.com/envoyproxy/envoy/blob/a18611c9a81a87b49470846eff1f62cc8507043b/source/extensions/filters/http/buffer/buffer_filter.cc#L100

Yeah I can add a section in the docs about it 👌

Reply:
Ultimately the issue is that the setup would still be a bit kludgy because of filter ordering, and the fact that the buffer filter can't be disabled independently for requests and responses. For Lyft's use case it'll be fine because we don't need to preserve content-length on responses.

So at the end of the day it might make sense to have this filter pay the price and do buffering itself. My proposal is to leave as is (with the new docs I added) to get this PR in, and then I can open an issue with a link to the commit that had buffering, it would just need to be updated a bit, and integration tests written (which shouldn't be hard at all). I can get to it as a low-pri.

Reply:
I think it's fine to document the limitation and we can revisit later.